axum-core/src/extract/mod.rs

```diff
@@ -65,256 +72,91 @@ pub use self::from_ref::FromRef;
 /// [`http::Request<B>`]: http::Request
 /// [`axum::extract`]: https://docs.rs/axum/latest/axum/extract/index.html
 #[async_trait]
-pub trait FromRequest<S, B>: Sized {
+pub trait FromRequestMut<S, B>: Sized {
     /// If the extractor fails it'll use this "rejection" type. A rejection is
     /// a kind of error that can be converted into a response.
     type Rejection: IntoResponse;
 
     /// Perform the extraction.
-    async fn from_request(req: &mut RequestParts<S, B>) -> Result<Self, Self::Rejection>;
+    async fn from_request_mut(req: &mut Request<B>, state: &S) -> Result<Self, Self::Rejection>;
 }
 
-/// The type used with [`FromRequest`] to extract data from requests.
-///
-/// Has several convenience methods for getting owned parts of the request.
-#[derive(Debug)]
-pub struct RequestParts<S, B> {
-    pub(crate) state: Arc<S>,
-    method: Method,
-    uri: Uri,
-    version: Version,
-    headers: HeaderMap,
-    extensions: Extensions,
-    body: Option<B>,
-}
+#[async_trait]
+pub trait FromRequestOnce<S, B, M = Once>: Sized {
```

Comment:
Oh so if you add `M` then it just works? Cool! :D

Reply:
Yeah, the trick is that a `FromRequestOnce<S, B, Once>` impl can't possibly overlap with a `FromRequestOnce<S, B, Mut>` impl. It could also be `FromRequestOnce<S, B, const REALLY_ONCE: bool = true>` with the blanket impl being for `FromRequestOnce<S, B, false>`, but then `Handler` also needs that const-generic param.